plotter/contour.go

```diff
@@ -0,0 +1,425 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:generate ./list
+
+package plotter
+
+import (
+	"math"
+	"sort"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/palette"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+)
+
+// Contour implements the Plotter interface, drawing
+// a contour plot of the values in the GridXYZ field. The
+// order of rows is from high index to low index down.
+type Contour struct {
+	GridXYZ GridXYZ
+
+	// Levels describes the contour heights to plot.
+	Levels []float64
+
+	// LineStyle is the style of the contour lines.
+	draw.LineStyle
+
+	// Palette is the color palette used to render
+	// the heat map.
+	Palette palette.Palette
+
+	// Min and Max define the dynamic range of the
+	// heat map.
+	Min, Max float64
+}
+
+// NewContour creates as new contour plotter for the given data,
```

Comment:
Calculating percentiles is quite costly in terms of allocation (all elements need to be allocated again in a slice to do the sort), but it can be done easily. What default for the chosen contours: 5, 25, 50, 75, 95?